src/system_metrics/service_mongoose_system_metrics.erl

```diff
@@ -0,0 +1,67 @@
+-module(service_mongoose_system_metrics).
+-author('jan.ciesla@erlang-solutions.com').
+
+-behaviour(gen_server).
+
+-define(DEFAULT_INITIAL_REPORT, 5 * 60 * 1000).     % 5 min intial report
+-define(DEFAULT_REPORT_AFTER, 3 * 60 * 60 * 1000).  % 3 hours periodic report
+
+-include("mongoose.hrl").
+
+-export([start/1, stop/0]).
+-export([start_link/1,
+         init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2]).
+
+-type system_metrics_state() :: pos_integer().
+
+-spec start(proplists:proplist()) -> {ok, pid()}.
+start(Args) ->
+    Spec = {?MODULE, {?MODULE, start_link, [Args]}, temporary, brutal_kill, worker, [?MODULE]},
+    {ok, _} = ejabberd_sup:start_child(Spec).
+
+-spec stop() -> ok.
+stop() ->
+    ejabberd_sup:stop_child(?MODULE).
+
+-spec start_link(proplists:proplist()) -> {ok, pid()}.
+start_link(Args) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, Args, []).
+
+-spec init(proplists:proplist()) -> {ok, system_metrics_state()}.
+init(Args) ->
+    InitialReport = proplists:get_value(initial_report, Args, ?DEFAULT_INITIAL_REPORT),
+    ReportAfter = proplists:get_value(report_after, Args, ?DEFAULT_REPORT_AFTER),
+    load_client_id(),
+    erlang:send_after(InitialReport, self(), spawn_gatherer),
+    {ok, ReportAfter}.
+    
+handle_info(spawn_gatherer, ReportAfter) ->
+    spawn(mongoose_system_metrics_gatherer, gather, []),
+    erlang:send_after(ReportAfter, self(), spawn_gatherer),
+    {noreply, ReportAfter};
+handle_info(_Message, _State) ->
+    ok.
+
+% %%-----------------------------------------
+% %% Helpers
+% %%-----------------------------------------
+
+load_client_id() ->
```

Comment:
I think we could roll back to the previous approach with ram only mnesia table so that it get's replicated to other nodes in the cluster.
Then in order to survive total cluster restart (when all nodes go down) we can also safe the value to a file and read. With this approach we need to remember about synchronization I suggest the following:
When we read the value from the mnesia table and
1. There is no value but there is a file - we read the value from file
2. There is no value and there is no file - we create the file, generate the value and write to both mnesia and file.
3. There is value - we write the value to the local file.

I know this is not free from race conditions. On the other hand, adding nodes to a cluster doesn't happen very often and usually one node is started after the other, especially when we set it up from scratch.

Reply:
Sure, rolled back to ram_copies for mnesia table.

Reply:
Let's move creating a file to another story/task as it might require more planning and testing.